examples/chat/src/main.rs

```diff
@@ -1,37 +1,88 @@
+const SESSION: &str = "via-chat-session";
+
+type Request = via::Request<Chat>;
+type Next = via::Next<Chat>;
 
 #[tokio::main]
 async fn main() -> Result<ExitCode, Error> {
-    let mut app = App::new(Chat::new());
+    dotenvy::dotenv()?;
+
+    let mut app = {
+        let pool = chat::establish_pg_connection().await;
+        let secret = chat::load_session_secret();
+
+        App::new(Chat::new(pool, secret))
+    };
+
+    app.uses(Cookies::new().allow(SESSION));
+    app.uses(Auth::new(SESSION));
+
+    app.route("/").to(via::get(routes::home));
+
+    let mut api = app.route("/api");
+
+    api.uses(Rescue::with(util::error_sanitizer));
+
+    // Perform a websocket upgrade and start chatting.
+    api.route("/subscribe").to(ws::upgrade(routes::subscribe));
+
+    api.route("/threads").scope(|threads| {
+        let mut thread = {
+            let (collection, member) = via::rest!(routes::threads);
+
+            threads.route("/").to(collection);
+            threads.route("/:thread-id").to(member)
+        };
+
+        // 403 when the current user is not in the thread.
+        thread.uses(routes::threads::authorization);
+
+        thread.route("/messages").scope(|messages| {
+            let mut message = {
+                let (collection, member) = via::rest!(routes::messages);
+
+                messages.route("/").to(collection);
+                messages.route("/:message-id").to(member)
+            };
+
+            message.route("/reactions").scope(|reactions| {
+                let (collection, member) = via::rest!(routes::reactions);
 
-    app.route("/").to(via::get(async |_, _| {
-        Response::build()
-            .header(header::CONTENT_SECURITY_POLICY, CSP)
-            .text("Chat Example frontend coming soon!")
-    }));
+                reactions.route("/").to(collection);
+                reactions.route("/:reaction-id").to(member);
+            });
+        });
 
-    // Define the router namespace for our chat API.
-    {
-        let mut room = app.route("/chat/:room");
+        // Access control is defined by the owner of the thread.
+        thread.route("/users").scope(|users| {
+            use routes::threads::{add, remove};
+            users.route("/").to(via::post(add).delete(remove));
+        });
+    });
 
-        // list all the messages in the room.
-        room.route("/").to(via::get(room::index));
+    api.route("/users").scope(|users| {
+        use routes::users::{login, logout};
 
-        // websocket to read / write messages.
-        room.route("/join").to(via::ws(room::join));
+        let (collection, member) = via::rest!(routes::users);
 
-        // show the message with the provided index.
-        room.route("/:index").to(via::get(room::show));
-    }
+        users.route("/").to(collection);
+        users.route("/:id").to(member);
+        users.route("/login").to(via::post(login));
+        users.route("/logout").to(via::post(logout));
```

Comment:
These are the only 2 routes that require an allocation to resolve. It's because there is a dynamic pattern adjacent to sibling routes with a static pattern.

From a router allocation perspective, this gives REST the same routing allocation behavior as GraphQL.